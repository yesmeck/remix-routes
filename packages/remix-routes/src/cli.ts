#!/usr/bin/env node
import meow from 'meow';
import * as fs from 'fs';
import * as path from 'path';
import chokidar from 'chokidar';
import type { ConfigRoute } from '@remix-run/dev/dist/config/routes';
import mkdirp from 'mkdirp';

let readConfig: typeof import('@remix-run/dev/dist/config').readConfig;

try {
  readConfig = require('@remix-run/dev/config').readConfig;
} catch (e) {
  readConfig = require('@remix-run/dev/dist/config').readConfig;
}

const helpText = `
Usage
$ remix-routes

Options
--watch, -w  Watch for routes changes
--outputDirPath, -o Specify the output path for "remix-routes.d.ts". Defaults to "./node_modules" if arg is not given.
`;

const DEFAULT_OUTPUT_DIR_PATH = './node_modules'

const cli = meow(helpText, {
  flags: {
    watch: {
      type: 'boolean',
      alias: 'w',
    },
    outputDirPath: {
      type: 'string',
      alias: 'o',
      default: DEFAULT_OUTPUT_DIR_PATH,
    }
  },
});

type RoutesInfo = Record<string, {
  fileName: string;
  params: string[];
}>

async function buildHelpers(remixRoot: string): Promise<RoutesInfo> {
  const config = await readConfig(remixRoot);
  const routesInfo: RoutesInfo = {};
  const handleRoutesRecursive = (
    parentId?: string,
    parentPath: ConfigRoute[] = [],
  ) => {
    let routes = Object.values(config.routes).filter(
      (route) => route.parentId === parentId,
    );
    routes.forEach((route) => {
      let currentPath = parentPath;
      if (route.path) {
        currentPath = [...currentPath, route];
        const fullPath = currentPath.reduce(
          (acc, curr) => [acc, trimSlash(curr.path!)].join('/'),
          '',
        );
        const paramsNames = parse(currentPath);
        routesInfo[fullPath] = {
          fileName: route.file,
          params: paramsNames
        };
      } else if (route.id === 'root') {
        routesInfo['/'] = {
          fileName: route.file,
          params: [],
        };
      }
      handleRoutesRecursive(route.id, currentPath);
    });
  };
  handleRoutesRecursive();
  return routesInfo;
}

export async function build(remixRoot: string, outputDirPath: string) {
  const routesInfo = await buildHelpers(remixRoot);
  generate(routesInfo, remixRoot, outputDirPath);
}

function watch(remixRoot: string, outputDirPath: string) {
  build(remixRoot, outputDirPath);
  chokidar
    .watch([
      path.join(remixRoot, 'app/routes/**/*.{ts,tsx}'),
      path.join(remixRoot, 'remix.config.js'),
    ])
    .on('change', () => {
      build(remixRoot, outputDirPath);
    });
  console.log('Watching for routes changes...');
}

function generate(routesInfo: RoutesInfo, remixRoot: string, outputDirPath: string) {
  const tsCode =
    [
      `
declare module "remix-routes" {

type IsAny<T> = (
  unknown extends T
    ? [keyof T] extends [never] ? false : true
    : false
);
type URLSearchParamsInit = string | string[][] | Record<string, string> | URLSearchParams;
type Query<T> = IsAny<T> extends true ? URLSearchParamsInit : T;
      `,
      generateRouteDefinition(routesInfo),
      generatePathDefinition(routesInfo),
      generateParamsDefinition(routesInfo),
      "}",
    ].join('\n\n') + '\n\n';

  const outputPath = path.join(
    remixRoot,
    outputDirPath,
  );

  if (!fs.existsSync(outputPath)) {
    mkdirp.sync(outputPath);
  }
  fs.writeFileSync(path.join(outputPath, 'remix-routes.d.ts'), tsCode);
}

function generateRouteDefinition(routesInfo: RoutesInfo) {
  const code: string[] = ['export interface Routes {'];
  Object.entries(routesInfo).forEach(([route, { fileName, params }]) => {
    const lines = [`  "${route}": {`];
    const paramsType = params.map(
      (param) => `${param}: string | number`,
    );
    lines.push(`    params: { ${paramsType.join('; ')} },`);
    lines.push(`    query: Query<import('../app/${fileName.replace(/\.tsx?$/, '')}').SearchParams>,`)
    lines.push('  };')
    code.push(lines.join('\n'));
  });
  code.push("}");
  return code.join('\n');
}

function generatePathDefinition(routesInfo: RoutesInfo) {
  const code: string[] = [];
  Object.entries(routesInfo).forEach(([route, { fileName, params }]) => {
    const lines = ['export declare function $path('];
    lines.push(`  route: ${JSON.stringify(route)},`);
    if (params.length > 0) {
      lines.push(`  params: Routes["${route}"]["params"],`);
    }
    lines.push(
      `  ...query: [Routes["${route}"]["query"]]`,
    );
    lines.push(`): string;`);
    code.push(lines.join('\n'));
  });
  return code.join('\n');
}

function generateParamsDefinition(routesInfo: RoutesInfo) {
  const routes = Object.entries(routesInfo);

  // $params helper makes sense only for routes with params.
  const routesWithParams = routes.filter(
    ([_, { params }]) => params.length > 0,
  );

  const code = routesWithParams.map(([route, { params }]) => {
    const lines: string[] = [];

    lines.push(`export declare function $params(`);
    lines.push(`  route: ${JSON.stringify(route)},`);
    lines.push(`  params: { readonly [key: string]: string | undefined }`);
    lines.push(`): {`);
    lines.push(
      params.map((param) => `  ${param}: string`).join(',\n'),
    );
    lines.push(`};`);

    return lines.join('\n');
  });

  return code.join('\n');
}

function parse(routes: ConfigRoute[]) {
  const paramNames: string[] = [];
  routes.forEach((route) => {
    return (
      route.path &&
      paramNames.push(
        ...route.path
          .split('/')
          .filter((seg) => seg.startsWith(':'))
          .map((param) => param.replace(':', '')),
      )
    );
  });
  return paramNames;
}

if (require.main === module) {
  (async function () {
    const remixRoot = process.env.REMIX_ROOT || process.cwd();

    if (cli.flags.watch) {
      watch(remixRoot, cli.flags.outputDirPath);
    } else {
      build(remixRoot, cli.flags.outputDirPath);
    }
  })();
}

function trimSlash(path: string) {
  return path.replace(/\/+$/, '');
}
