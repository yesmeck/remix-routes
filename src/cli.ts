#!/usr/bin/env node
import meow from 'meow';
import * as fs from 'fs';
import * as path from 'path';
import chokidar from 'chokidar';
import type { ConfigRoute } from '@remix-run/dev/dist/config/routes';
import mkdirp from 'mkdirp';

let readConfig: typeof import('@remix-run/dev/dist/config').readConfig;

try {
  readConfig = require('@remix-run/dev/config').readConfig;
} catch (e) {
  readConfig = require('@remix-run/dev/dist/config').readConfig;
}

const helpText = `
Usage
$ remix-routes

Options
--watch, -w  Watch for routes changes
`;

const cli = meow(helpText, {
  flags: {
    watch: {
      type: 'boolean',
      alias: 'w',
    },
  },
});

type RoutesInfo = Record<string, string[]>;

async function buildHelpers(remixRoot: string): Promise<RoutesInfo> {
  const config = await readConfig(remixRoot);
  const routesInfo: RoutesInfo = {};
  const handleRoutesRecursive = (
    parentId?: string,
    parentPath: ConfigRoute[] = [],
  ) => {
    let routes = Object.values(config.routes).filter(
      (route) => route.parentId === parentId,
    );
    routes.forEach((route) => {
      let currentPath = parentPath;
      if (route.path) {
        currentPath = [...currentPath, route];
        const fullPath = currentPath.reduce(
          (acc, curr) => [acc, trimSlash(curr.path!)].join('/'),
          '',
        );
        const paramsNames = parse(currentPath);
        routesInfo[fullPath] = paramsNames;
      }
      handleRoutesRecursive(route.id, currentPath);
    });
  };
  handleRoutesRecursive();
  return routesInfo;
}

export async function build(remixRoot: string) {
  const routesInfo = await buildHelpers(remixRoot);
  generate(routesInfo);
}

function watch(remixRoot: string) {
  build(remixRoot);
  chokidar
    .watch([
      path.join(remixRoot, 'app/routes/**/*.{ts,tsx}'),
      path.join(remixRoot, 'remix.config.js'),
    ])
    .on('change', () => {
      build(remixRoot);
    });
  console.log('Watching for routes changes...');
}

function generate(routesInfo: RoutesInfo) {
  const jsCode = generateHelpers(routesInfo);
  const tsCode =
    [
      generatePathDefinition(routesInfo),
      generateParamsDefinition(routesInfo),
    ].join('\n\n') + '\n\n';

  const outputPath = path.join(
    process.cwd(),
    'node_modules',
    'remix-routes',
  );

  if (!fs.existsSync(outputPath)) {
    mkdirp.sync(outputPath);
  }
  fs.writeFileSync(path.join(outputPath, 'index.js'), jsCode);
  fs.writeFileSync(path.join(outputPath, 'index.d.ts'), tsCode);
  fs.writeFileSync(
    path.join(outputPath, 'package.json'),
    JSON.stringify({
      name: 'remix-routes',
      main: 'index.js',
    }),
  );
}

function generateHelpers(routesInfo: RoutesInfo) {
  const routes = Object.entries(routesInfo).reduce(
    (acc: Record<string, string[]>, [route, params]) => {
      if (params.length > 0) {
        acc[route] = params;
      }
      return acc;
    },
    {},
  );
  const helper = fs.readFileSync(path.resolve(__dirname, '../helper.js'), 'utf-8')
  return helper.replace('// __routes__', `const routes = ${JSON.stringify(routes, null, 2)}`);
}

function generatePathDefinition(routesInfo: RoutesInfo) {
  const code: string[] = [];
  Object.entries({
    '/': [],
    ...routesInfo,
  }).forEach(([route, paramsNames]) => {
    const lines = ['export declare function $path('];
    lines.push(`  route: ${JSON.stringify(route)},`);
    if (paramsNames.length > 0) {
      const paramsType = paramsNames.map(
        (paramName) => `${paramName}: string | number`,
      );
      lines.push(`  params: { ${paramsType.join('; ')} },`);
    }
    lines.push(
      `  query?: string | string[][] | Record<string, string> | URLSearchParams`,
    );
    lines.push(`): string;`);
    code.push(lines.join('\n'));
  });
  return code.join('\n');
}

function generateParamsDefinition(routesInfo: RoutesInfo) {
  const routes = Object.entries(routesInfo);

  // $params helper makes sense only for routes with params.
  const routesWithParams = routes.filter(
    ([_, paramsNames]) => paramsNames.length > 0,
  );

  const code = routesWithParams.map(([route, paramsNames]) => {
    const lines: string[] = [];

    lines.push(`export declare function $params(`);
    lines.push(`  route: ${JSON.stringify(route)},`);
    lines.push(`  params: { readonly [key: string]: string | undefined }`);
    lines.push(`): {`);
    lines.push(
      paramsNames.map((paramName) => `  ${paramName}: string`).join(',\n'),
    );
    lines.push(`};`);

    return lines.join('\n');
  });

  return code.join('\n');
}

function parse(routes: ConfigRoute[]) {
  const paramNames: string[] = [];
  routes.forEach((route) => {
    return (
      route.path &&
      paramNames.push(
        ...route.path
          .split('/')
          .filter((seg) => seg.startsWith(':'))
          .map((param) => param.replace(':', '')),
      )
    );
  });
  return paramNames;
}

if (require.main === module) {
  (async function () {
    const remixRoot = process.env.REMIX_ROOT || process.cwd();

    if (cli.flags.watch) {
      watch(remixRoot);
    } else {
      build(remixRoot);
    }
  })();
}

function trimSlash(path: string) {
  return path.replace(/\/+$/, '');
}
