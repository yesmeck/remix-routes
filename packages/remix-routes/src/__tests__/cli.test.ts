import { test, expect } from 'vitest';
import * as fs from 'fs';
import * as path from 'path';
import { build } from '../cli';

test('build', async () => {
  await build(path.resolve(__dirname, '../../fixture'));
  expect(
    fs.readFileSync(path.resolve(__dirname, '../../fixture/node_modules/remix-routes.d.ts'), 'utf8'),
  ).toMatchSnapshot();
});
