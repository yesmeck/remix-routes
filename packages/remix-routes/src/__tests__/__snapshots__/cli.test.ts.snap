// Vitest Snapshot v1

exports[`build 1`] = `
"declare module \\"remix-routes\\" {
  type IsAny<T> = (
    unknown extends T
      ? [keyof T] extends [never] ? false : true
      : false
  );
  type URLSearchParamsInit = string | string[][] | Record<string, string> | URLSearchParams;
  type ExportedQuery<T> = IsAny<T> extends true ? URLSearchParamsInit : T;
  type Query<T> = IsAny<T> extends true ? [URLSearchParamsInit?] : [T];

  export interface Routes {
  
    \\"/\\": {
      params: {
      
      },
      query: ExportedQuery<import('../app/root').SearchParams>,
    };
  
    \\"/chats/:season/:episode/:slug\\": {
      params: {
      
        season: string | number;
      
        episode: string | number;
      
        slug: string | number;
      
      },
      query: ExportedQuery<import('../app/routes/chats.$season.$episode.$slug').SearchParams>,
    };
  
    \\"/chats/:season/:episode\\": {
      params: {
      
        season: string | number;
      
        episode: string | number;
      
      },
      query: ExportedQuery<import('../app/routes/chats.$season.$episode').SearchParams>,
    };
  
    \\"/people/:personId\\": {
      params: {
      
        personId: string | number;
      
      },
      query: ExportedQuery<import('../app/routes/people/$personId').SearchParams>,
    };
  
    \\"/people/:personId/:planId/remove-plan\\": {
      params: {
      
        personId: string | number;
      
        planId: string | number;
      
      },
      query: ExportedQuery<import('../app/routes/people/$personId/$planId/remove-plan').SearchParams>,
    };
  
    \\"/blog/rss.xml\\": {
      params: {
      
      },
      query: ExportedQuery<import('../app/routes/blog.rss[.]xml').SearchParams>,
    };
  
    \\"/:lang?/about\\": {
      params: {
      
        lang?: string | number;
      
      },
      query: ExportedQuery<import('../app/routes/($lang)/about').SearchParams>,
    };
  
    \\"/posts/delete\\": {
      params: {
      
      },
      query: ExportedQuery<import('../app/routes/posts/delete').SearchParams>,
    };
  
    \\"/auth\\": {
      params: {
      
      },
      query: ExportedQuery<import('../app/routes/auth/__auth').SearchParams>,
    };
  
    \\"/auth/login\\": {
      params: {
      
      },
      query: ExportedQuery<import('../app/routes/auth/__auth/login').SearchParams>,
    };
  
    \\"/posts\\": {
      params: {
      
      },
      query: ExportedQuery<import('../app/routes/posts/index').SearchParams>,
    };
  
    \\"/posts/:id\\": {
      params: {
      
        id: string | number;
      
      },
      query: ExportedQuery<import('../app/routes/posts/$id').SearchParams>,
    };
  
    \\"/s/:query\\": {
      params: {
      
        query: string | number;
      
      },
      query: ExportedQuery<import('../app/routes/s.$query').SearchParams>,
    };
  
    \\"/credits\\": {
      params: {
      
      },
      query: ExportedQuery<import('../app/routes/credits').SearchParams>,
    };
  
    \\"/posts/*\\": {
      params: {
      
        \\"*\\": string | number;
      
      },
      query: ExportedQuery<import('../app/routes/posts/$').SearchParams>,
    };
  
    \\"/admin\\": {
      params: {
      
      },
      query: ExportedQuery<import('../app/routes/admin').SearchParams>,
    };
  
    \\"/admin/episodes\\": {
      params: {
      
      },
      query: ExportedQuery<import('../app/routes/admin/episodes/index').SearchParams>,
    };
  
    \\"/admin/episodes/:id\\": {
      params: {
      
        id: string | number;
      
      },
      query: ExportedQuery<import('../app/routes/admin/episodes/$id').SearchParams>,
    };
  
    \\"/admin/episodes/:id/comments\\": {
      params: {
      
        id: string | number;
      
      },
      query: ExportedQuery<import('../app/routes/admin/episodes/$id/comments').SearchParams>,
    };
  
    \\"/admin/episodes/new\\": {
      params: {
      
      },
      query: ExportedQuery<import('../app/routes/admin/episodes/new').SearchParams>,
    };
  
    \\"/jokes\\": {
      params: {
      
      },
      query: ExportedQuery<import('../app/routes/jokes').SearchParams>,
    };
  
    \\"/jokes/:jokeId\\": {
      params: {
      
        jokeId: string | number;
      
      },
      query: ExportedQuery<import('../app/routes/jokes/$jokeId').SearchParams>,
    };
  
  }

  type RoutesWithParams = Pick<
    Routes,
    {
      [K in keyof Routes]: Routes[K][\\"params\\"] extends Record<string, never> ? never : K
    }[keyof Routes]
  >;

  export declare function $path<
    Route extends keyof Routes,
    Rest extends {
      params: Routes[Route][\\"params\\"];
      query?: Routes[Route][\\"query\\"];
    }
  >(
    ...args: Rest[\\"params\\"] extends Record<string, never>
      ? [route: Route, query?: Rest[\\"query\\"]]
      : [route: Route, params: Rest[\\"params\\"], query?: Rest[\\"query\\"]]
  ): string;

  export declare function $params<
    Route extends keyof RoutesWithParams,
    Params extends RoutesWithParams[Route][\\"params\\"]
  >(
      route: Route,
      params: { readonly [key: string]: string | undefined }
  ): {[K in keyof Params]: string};
}"
`;
